File: bnovate/bnovate/report/ic_profit_and_loss_statement/ic_profit_and_loss_statement.js
// Copyright (c) 2024, libracore and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["IC Profit And Loss Statement"] = {
    "filters": [
        {
            "fieldname":"from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": new Date((new Date()).getFullYear(), 0, 1),
            "reqd": 1
        },
        {
            "fieldname":"to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": new Date((new Date()).getFullYear(), 11, 31),
            "reqd": 1
        },
        {
            "fieldname":"eur_exchange_rate",
            "label": __("EUR Exchange Rate"),
            "fieldtype": "Float",
            "default": 1,
            "reqd": 1
        },
        {
            "fieldname":"gbp_exchange_rate",
            "label": __("GBP Exchange Rate"),
            "fieldtype": "Float",
            "default": 1,
            "reqd": 1
        }
    ],
    "onload": (report) => {
        frappe.db.get_value(
            "Currency Exchange", 
            filters={'from_currency': 'GBP', 'to_currency': 'CHF'}, 
            fields=['exchange_rate']
        ).then( function (response) {
            frappe.query_report.get_filter('gbp_exchange_rate').set_value(response.message.exchange_rate);
        });
        frappe.db.get_value(
            "Currency Exchange", 
            filters={'from_currency': 'EUR', 'to_currency': 'CHF'}, 
            fields=['exchange_rate']
        ).then( function (response) {
            frappe.query_report.get_filter('eur_exchange_rate').set_value(response.message.exchange_rate);
        });
    }
};

File: bnovate/bnovate/report/ic_profit_and_loss_statement/ic_profit_and_loss_statement.py
# Copyright (c) 2024, libracore and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    # add head column
    columns = [
        {'fieldname': 'account_group', 'label': 'Account Group', 'fieldtype': 'Data', 'width': 200}
    ]
    # add column per company
    companies = frappe.get_all("Company", fields=['name', 'abbr'])
    for c in companies:
        columns.append({
            'fieldname': '{0}'.format(c['abbr']),
            'label': c['abbr'],
            'fieldtype': 'Currency',
            'width': 120
        })
    
    columns.append({
        'fieldname': 'total',
        'label': 'Total',
        'fieldtype': 'Currency',
        'width': 120
    })
    
    return columns
    
def get_data(filters=None):
    data = []
    
    # fetch all account groups
    node_account_groups = frappe.db.sql("""
        SELECT `name`, `is_group`, `level`
        FROM `tabIC Account Group`
        WHERE
            `reporting_type` IN ("Income", "Expense")
            AND (`parent_group` IS NULL OR `parent_group` = "")
        ORDER BY `title` ASC;
        """, as_dict=True)
    
    # build group tree
    account_group_tree = []
    for node in node_account_groups:
        account_group_tree.append(recurse_account_group(node, filters))
    
    # expand tree into a list
    data = expand_node(data, account_group_tree)
    
    # add totals per row
    companies = frappe.get_all("Company", fields=['name', 'abbr'])
    for d in data:
        row_total = 0
        for c in companies:
            row_total += d['{0}'.format(c['abbr'])]
        d['total'] = row_total
        
    return data

def expand_node(data, children):
    for d in children:
        _data = {
            'account_group': d['name'],
            'indent': d['level']
        }
        
        for k,v in d.items():
            if k not in ['name', 'is_group', 'level', 'children']:
                _data[k] = v
                        
        data.append(_data)
        
        if 'children' in d:
            data = expand_node(data, d['children'])
    return data
        
def recurse_account_group(node, filters):
    account_group = {
        'name': node['name'], 
        'is_group': node['is_group'],
        'level': node['level']
    }
    
    if node['is_group']:
        account_group['children'] = []
        children = frappe.get_all(
            "IC Account Group", 
            filters={'parent_group': node['name']}, 
            fields=['name', 'is_group', 'level'],
            order_by='title ASC'
        )
        for child in children:
            _child = recurse_account_group(child, filters)
            account_group['children'].append(_child)
            for k,v in _child.items():
                if k not in ['name', 'is_group', 'level', 'children']:
                    if k not in account_group:
                        account_group[k] = (v or 0)
                    else:
                        account_group[k] += (v or 0)
            
    else:
        account_group.update(get_balance_for_account_group(node['name'], filters))
    
    return account_group
        
def get_balance_for_account_group(account_group, filters):
    group_doc = frappe.get_doc("IC Account Group", account_group)
    group_balances = {}
    for account in group_doc.accounts:
        if account.abbr not in group_balances:
            group_balances[account.abbr] = 0
        _balance = get_balance(account.account, filters.get('from_date'), filters.get('to_date'))
        if account.currency == "GBP":
            _balance = _balance * filters.get('gbp_exchange_rate')
        elif account.currency == "EUR":
            _balance = _balance * filters.get('eur_exchange_rate')
        group_balances[account.abbr] += _balance
    
    return group_balances
    
def get_balance(account, from_date, to_date):
    balance = frappe.db.sql("""
        SELECT IFNULL(SUM(`credit`), 0) - IFNULL(SUM(`debit`), 0) AS `balance`
        FROM `tabGL Entry`
        WHERE
            `account` = "{account}"
            AND `posting_date` BETWEEN "{from_date}" AND "{to_date}"
            AND `voucher_type` != "Period Closing Voucher"
        ;""".format(account=account, from_date=from_date, to_date=to_date), as_dict=True)
    if len(balance) > 0:
        return balance[0]['balance']
    else:
        return 0
        
